VTAdmin Intro Blogpost (#1266)
vitessio/website: content/en/blog/2022-11-28-vtadmin-intro.md

---
author: 'Frances Thai'
date: 2022-11-28
slug: '2022-11-28-vtadmin-intro'
tags: ['Vitess','VTAdmin','vtctld','vtctld2']
title: 'Introducing VTAdmin'
description: "Vitess's cluster management API and UI"
---

[VTAdmin](https://vitess.io/docs/reference/vtadmin/) is now generally available for use! VTAdmin provides both a web client and API for managing multiple Vitess clusters, and is the successor to the now-deprecated UI for [vtctld](https://vitess.io/docs/reference/programs/vtctld/).

## What is VTAdmin?
VTAdmin is made up of two components: 
- VTAdmin API: An HTTP(S) and gRPC API server
- VTAdmin Web: A React + Typescript web client built with [Create React App](https://create-react-app.dev/)

## What can you do with VTAdmin?
VTAdmin can do everything the old vtctld2 UI could do, now that the [vtctld2 parity project](https://github.com/vitessio/vitess/projects/13) has been completed. For a complete list of supported API methods, refer [here](https://github.com/vitessio/vitess/blob/main/go/vt/vtadmin/api.go#L332).

The following are just a few examples of what VTAdmin can do!

> ### Tablet Management

Reviewer: To be consistent with the rest of the website we should use `VTTablet` everywhere.

> VTAdmin provides a variety of Tablet management tools - from starting and stopping replication, setting to read/write, reparenting, deleting, pinging, refreshing Tablets to experimental features like tablet and VReplication QPS. 

Reviewer: Does tablet and VReplication QPS mean "VTTablet QPS and VReplication QPS"?
It was not obvious how that should be parsed, but making that "VTTablet and VReplication QPS" is probably good enough.

Author: Yes! I'll update it to that 😄

<img src="/files/2022-11-28-vtadmin-intro/tablets.gif" alt="GIF of tablets features in VTAdmin Web"/>

_Note: To use experimental features, make sure to set `REACT_APP_ENABLE_EXPERIMENTAL_TABLET_DEBUG_VARS` in VTAdmin Web and `--http-tablet-url-tmpl` in VTAdmin API, as experimental tablet features work by making an HTTP call to the VTTablets._

### Keyspace Management

> In VTAdmin, keyspace actions like validate keyspace/schema/version, reloading schemas, rebuilding keyspace graphs and cells, and creating new shards are made easy.

Reviewer: ```suggestion
In VTAdmin, keyspace actions like validating keyspace/schema/version, reloading schemas, rebuilding keyspace graphs and cells, and creating new shards are made easy.
```

<img src="/files/2022-11-28-vtadmin-intro/keyspaces.gif" alt="GIF of keyspace features in VTAdmin Web"/>

> ## Workflow Management

Reviewer: We should probably make it clear either in the heading or in the paragraph text that these are VReplication workflows. Maybe the paragraph is a better place for it.

Author: I'll add it to the paragraph! should we update the UI to make that clearer as well?

Reviewer: I think UI is fine, it's sort of self-explanatory when you see them.

Author: sounds good!

> VTAdmin also allows you to view all your workflows and monitor workflow streams.

Reviewer: No need to say "also".

<img src="/files/2022-11-28-vtadmin-intro/workflows.gif" alt="GIF of workflow features in VTAdmin Web"/>

### Topology

> The old topology browser in vtctld2 has also been reimagined into a graph-traversal UI, that allows you to explore topology across single and multi-cluster deployments.

Reviewer: ```suggestion
The old topology browser in vtctld2 has been reimagined into a graph-traversal UI, which allows you to explore the full topology across single and multi-cluster deployments.
```

Reviewer: Source: https://www.diffen.com/difference/That_vs_Which 😄

> <img src="/files/2022-11-28-vtadmin-intro/topology.gif" alt="GIF of topology in VTAdmin Web"/>

Reviewer: We should use names like Vitess Cluster A and Vitess Cluster B (or test/dev/staging and prod) for each of these clusters.
Using AWS region names is misleading because a single Vitess cluster can span multiple regions.

Reviewer: Since we use local and prod in a later image, those maybe good names to use here as well.

Author: The updated topology image uses the local + prod examples, does that work?

Reviewer: Sorry this comment was in the wrong place (I was switching back and forth between preview and PR).
I'm talking about the `architecture.png` file. It has us-east-1 and ap-south-1. Topology image is good.

Author: Ohhh gotcha. I'll have to remake that in illustrator then!

### Cluster configuration and discovery
VTAdmin manages to be stateless because it mostly proxies queries to VTGates and Vtctlds within Vitess clusters. It is able to do this through **cluster discovery**, the mechanism by which addresses for VTGates and Vtctlds are discovered.

Discovery is specified as a part of [cluster configuration](https://github.com/vitessio/vitess/blob/main/doc/vtadmin/clusters.yaml). Cluster configuration is passed as the command line argument `--cluster` to VTAdmin API like so:
```bash
vtadmin \
  --addr ":14200" \
  --http-origin "http://localhost:14201" \
  --http-tablet-url-tmpl "http://{{ .Tablet.Hostname }}:15{{ .Tablet.Alias.Uid }}" \
  --tracer "opentracing-jaeger" \
  --grpc-tracing \
  --http-tracing \
  --logtostderr \
  --alsologtostderr \
  --rbac \
  --rbac-config="./vtadmin/rbac.yaml" \
  --cluster "id=local,name=local,discovery=staticfile,discovery-staticfile-path=./vtadmin/discovery.json,tablet-fqdn-tmpl={{ .Tablet.Hostname }}:15{{ .Tablet.Alias.Uid }}"
```
where, in this example, `discovery=staticfile` is specifying static file discovery.

VTAdmin API currently supports a few methods for discovery:

> #### Consul discovery

Reviewer: We discovered recently that Consul support was actually deprecated in Vitess 6.0.
So I'd like us to not write about Consul here. Instead let's add a section on Dynamic discovery after static discovery.

Author: On it! :)

                "hostname": "localhost:15991"
            }
        }
    ]
}
```

where `fqdn` specifies the address of the component's web UI, and `hostname` specifies the address of the component's gRPC server.

##### Multiple clusters
To specify multiple clusters, repeat the `--cluster` flag:
```bash
vtadmin \
  --addr ":14200" \
  --http-origin "http://localhost:14201" \
  --http-tablet-url-tmpl "http://{{ .Tablet.Hostname }}:15{{ .Tablet.Alias.Uid }}" \
  --tracer "opentracing-jaeger" \
  --grpc-tracing \
  --http-tracing \
  --logtostderr \
  --alsologtostderr \
  --rbac \
  --rbac-config="./vtadmin/rbac.yaml" \
  --cluster "id=local,name=local,discovery=staticfile,discovery-staticfile-path=./vtadmin/discovery-local.json,tablet-fqdn-tmpl={{ .Tablet.Hostname }}:15{{ .Tablet.Alias.Uid }}"
  --cluster "id=prod,name=prod,discovery=staticfile,discovery-staticfile-path=./vtadmin/discovery-prod.json,tablet-fqdn-tmpl={{ .Tablet.Hostname }}:15{{ .Tablet.Alias.Uid }}"
```

The above multi-cluster configuration would show up in VTAdmin Web as:

<img src="/files/2022-11-28-vtadmin-intro/multiclusters.png" alt="Multiple clusters on the /clusters page in VTAdmin"/>

> 

Reviewer: We should have a section here on how to get started, and state that VTAdmin is already integrated into both the local example and operator example (or rather the user guides), and link to those.

Author: There's a "How do I operate VTAdmin?" section that links to the articles on how to get started, and includes links to the local and operator examples. Should I paraphrase those here?

Reviewer: Oh I think I just missed that. That is good. I see that the operator's guide already links the local example user guide (Getting Started)
Ideally we should update "Getting Started" to demonstrate VTAdmin as a separate PR.

Author: that sounds good! I am happy to update that in a later PR.

## What's next?
There a number of things the team is excited to do next! Some of those things include:
- **Single component VTAdmin**: VTAdmin is currently deployed as two separate components: the Web client and the API server. We're working on packaging these up into a single component much like how the old vtctld2 UI was packaged with Vtctld.

> - **Adding VTOrc UI**: We'd also like to combine VTOrc management capabilities into VTAdmin, primarily [the VTOrc UI](https://vitess.io/docs/15.0/user-guides/configuration-basic/vtorc/#old-ui-removal-and-replacement). That means instead of having to access two web applications to manage your Vitess clusters, you'll be able to do it all from VTAdmin.

Reviewer: Since we have already deleted the VTOrc UI, let's delete the second sentence from this.
